posthog/temporal/workflows/squash_person_overrides.py

```diff
@@ -0,0 +1,256 @@
+    posthog_personoverride
+USING
+    deleted
+WHERE
+    team_id = %(team_id)s
+    AND old_person_id = deleted.id
+"""
+
+
+@dataclass
+class QueryInputs:
+    """Inputs for activities that run queries in the SquashPersonOverrides workflow."""
+
+    partition_id: str | None = None
+    to_override: dict[int, set[str]] = field(default_factory=dict)
+    database: str = "default"
+    join_table_name: str = "person_overrides_join"
+
+
+@activity.defn
+async def select_persons_to_squash(inputs: QueryInputs) -> dict[int, set[str]]:
+    """Select the persons we'll override to lock them in and safely delete afterwards
+
+    New overrides may come in while we are executing this workflow, so we need to
+    preemptively select which persons we are going to override to ignore any new ones
+    coming in.
+    """
+    to_override = sync_execute(
+        SELECT_PERSONS_TO_OVERRIDE_QUERY.format(database=inputs.database), {"partition_id": inputs.partition_id}
+    )
+
+    if not isinstance(to_override, list):
+        # Could return None if no results or int if this were an insert.
+        # Mostly to appease type checker
+        return {}
+
+    get_team_id = operator.itemgetter(0)
+
+    return {team_id: set(row[1] for row in rows) for team_id, rows in itertools.groupby(to_override, key=get_team_id)}
+
+
+@activity.defn
+async def squash_events_partition(inputs: QueryInputs):
+    """Execute the squash query for a given partition_id and persons to_override.
+
+    As ClickHouse doesn't support an UPDATE ... FROM statement ala PostgreSQL, we must
+    do this in 4 steps/queries:
+
+    1. Build a JOIN table AS person_overrides.
+    2. Populate it with the data we are using in the update.
+    3. Perform ALTER TABLE UPDATE using joinGet to query the JOIN table.
+    4. Clean up the JOIN table once done.
+    """
+    activity.logger.info("Creating JOIN table %s.%s", inputs.database, inputs.join_table_name)
+    sync_execute(CREATE_JOIN_TABLE_QUERY.format(database=inputs.database, table_name=inputs.join_table_name))
+
+    activity.logger.info("Populating JOIN table %s.%s", inputs.database, inputs.join_table_name)
+    sync_execute(
+        INSERT_INTO_JOIN_TABLE_QUERY.format(database=inputs.database, table_name=inputs.join_table_name),
+        {"partition_id": inputs.partition_id},
+    )
+
+    activity.logger.info("Executing squash query on partition %s", inputs.partition_id)
+    sync_execute(
+        SQUASH_EVENTS_QUERY.format(database=inputs.database, table_name=inputs.join_table_name),
+        {"partition_id": inputs.partition_id},
+    )
+
+    activity.logger.info("Cleaning up JOIN table %s.%s", inputs.database, inputs.join_table_name)
+    sync_execute(DROP_JOIN_TABLE_QUERY.format(database=inputs.database, table_name=inputs.join_table_name))
+
+
+@activity.defn
+async def delete_squashed_person_overrides_from_clickhouse(inputs: QueryInputs):
+    """Execute the query to delete persons in ClickHouse that have been squashed."""
```

Comment:
Didn't we have the plan, that we'll only squash the last N partitions, in which case we can't delete the squashed persons as they could be in older partitions we didn't update this time. So what we want to likely be doing is do a query to events table to verify that these old_ids are not there and only for the old_ids that aren't there we'll remove them from Postgres. This makes the merge at older than X time even more important.

Reply:
Funnily enough, I think this is an issue only in the presence of duplicates: there should only be one `old_person_id` mapping, as enforced by postgres, so it's safe to delete it once we are done with it as there cannot be another `old_person_id` mapping in a different partition. 

However, this is only true if we account for duplicates, as CH won't enforce uniqueness constraints. If a second mapping for the same `old_person_id` arrives, then it's safe to prefer the newer one, as it basically overwrites the previous mapping. So, what we can do is be duplicate-aware in our queries and prefer newer ones in the presence of duplicates. 

This is a nice example of what we should do in all queries.

Reply:
By duplicates you mean updates, right? As the old_person_id -> override_id then we just have a inc in the version and separate rows in CH, no?

One problem is that while we're doing this someone could be adding a new override for that old_person, so we need to make sure we only delete the old->override (as it was when we selected in the beginning, if it still exists and not touch it otherwise)

Reply:
When you say delete it you mean in CH? Postgres is already enforcing only one old -> override exists, so if a new one arrives it will ensure the old one is deleted by whoever inserts the new one.

For CH we handle this by deleting overrides with a lower `merged_at`. For Postgres, unfortunately we don't have a `merged_at` field or any timestamp field besides `oldest_event` which is not very reliable. So, I would just take both ids from the mapping + the version to decide what to delete.

Reply:
Yeah, I'm pushing a commit in a bit that is as specific as it gets when it comes to deleting from Postgres (we check all ids and version when deleting).

It's really hard to coordinate things between two databases, I wish we didn't have to delete from PG at all...